File: Leetcode/2089. Find Target Indices After Sorting Array/solution.rb
#linear solution
def target_indices(nums, target)
  nums.sort!
  output = []

  nums.each.with_index do |num, index|
    if num == target
      output.push(index)
    end
  end

  output
end

#Binary Search solution
def target_indices(nums, target)
  nums.sort!
  output = []

  low = 0
  high = nums.length - 1

  while low <= high
    mid = low + ((high - low) / 2)
    puts nums[mid]

    if nums[mid] < target
      low = mid + 1
    elsif nums[mid] > target
      high = mid - 1
    else
      output << mid
      low = mid - 1
      high = mid + 1
      break
    end
  end

  # check left of target
  while low >= 0 && nums[low] == target
    output << low
    low -= 1
  end

  # check right of target
  while high < nums.length && nums[high] == target
    output << high
    high += 1
  end

  output.sort!
end

File: Leetcode/2089. Find Target Indices After Sorting Array/solution.js
const targetIndices = (nums, target) => {
  const outPut = []
  const sortedNums = nums.sort((a, b) => a - b)

  let left = 0
  let right = sortedNums.length - 1

  while (left <= right) {
    mid = Math.floor(left + ((right - left) / 2))
    if (sortedNums[mid] < target) {
      left = mid + 1
    } else if (sortedNums[mid] > target) {
      right = mid - 1
    } else {
      outPut.push(mid)
      left = mid - 1
      right = mid + 1
      break
    }
  }

  while (left >= 0 && sortedNums[left] === target) {
    outPut.push(left)
    left --
  }

  while (right < sortedNums.length && sortedNums[right] === target) {
    outPut.push(right)
    right ++
  }

  return outPut.sort((a, b) => a - b)
}

console.log(targetIndices([1,2,5,2,3], 3))